docs/spec/consensus/checkvalidators.md

```diff
@@ -0,0 +1,224 @@
+3. $nh.bfttime < oh.bfttime+tp$
+4. $trust(oh)=true$
+
+### Some Maths
+
+**Observation 1.** If $oh.bfttime + tp > now$, we trust the old
+validator set $oh.V$.
+
+In the following let's assume oh is trusted and sufficiently new.
+
+**Definition 1.** Let $PA \subseteq oh.V$ be a *potential adversary* in $oh$, if
+the sum of the voting powers in PA is less than 1/3 of the voting
+powers in $oh.V$, that is,
+\[
+\sum_{(v,p) \in PA} p < 1/3 \sum_{(v,p) \in oh.V} p
+\]
+
+**Proposition 1.** The set of faulty processes
+\[oh.V \setminus \{(v,p): (v,p) \in oh.V \wedge correct(v,oh.bfttime + tp)\}\]
+is a potential adversary.
+
+*Proof.* By the LightClient invariant.
+
+**Definition 2.** Let the *unknown validators* UV be the validators that appear in $nh.V$ and not in
+$oh.V$, that is,
+\[
+UV = \{(v,p): (v,p) \in nh.V \wedge \nexists (v,x) \in oh.V \}.
+\]
+
+
+
+
+**Theorem 1.** If for all potential adversaries PA, in $nh$ the combined voting
+powers of PA and UV is less than a third of the total voting power, then in
+$nh$, more than 2/3 of the voting power is held by correct processes. Formally,
+if for all PA
+
+\[
+\sum_{(v,old) \in PA \wedge \\ (v,p) \in nh.V} p + \sum_{(v,p) \in UV} p < 1/3
+\sum_{(v,p) \in nh.V} p,
+\]
+then
+\[
+\sum_{(v,p) \in nh.V \wedge \\correct(v,oh.bfttime + tp)} p > 2/3 \sum_{(v,p) \in h.V} p
+\]
+
+*Proof.* By the definition of PA, Proposition 1, and the LightClient invariant.
+
+By Assumption 3, there is sufficient voting power to trust the new validator set. (And thus the validator set it signs in that block, for which the $tp$ starts at the $bfttime$ of the header).
+
+Below, we thus sketch a function that checks whether the premise of Theorem 1 holds. If the results is positive, we can trust $nh$, otherwise not.
+
+
+
+### An Algorithm
+
+In pseudo go...
+
+```go
+func CheckVS(oh, nh) bool {
+  if oh.bfttime + unbonding_period < now { // Observation 1
+    return false // old header was once trusted but it is expired
+  }
+
+  PAs := compute_all_PAs(oh)  // Definition 1
+  PAs := reduce (PAs)         // remove every PA that is a subset of another PA
+  UV := compute_UV(oh,nh)     // Definition 2
+
+  vpUV := votingpower(UV,nh)  // second sum in Theorem 1
+  vpNH := votingpower(nh.V,nh) // right hand side of premise of Theorem 1
+  vpMaxPA := maximumvotingpower(PAs,nh) // voting powers of all PA and big max
+
+  return vpMaxPA + vpUV < 1/3 * vpNH  // Theorem 1. It must be smaller for all
+                                        // so it must be smaller for the max
+}
+```  
+
+
+## Remarks
+
+**Remark.** Computing all PAs might be too expensive (all subsets of $oh.V$ that have a certain combined voting power in oh). Similarly, we then have to compute all voting powers of PAs in nh to get the maximum. This is disturbing, as right now, based on the examples, I expect that CheckVS will mostly return false, assuming that there are frequent changes in the validator sets. However, $oh.V=nh.V$ might be the common case.
```

Comment:
I *think* we can just calculate the intersection between `oh.V` and `nh.V` (a set of validator keys), sum the total stake, subtract largest-first until we reach `1/3` (maximally conservative, assume the largest-stake faulty nodes intersected), and check if the resulting total is `>= 2/3` of stake at `nh.V`. Should be double-checked, but that algorithm should be linear time.

Reply:
Not sure I understand well your proposal, but I see two issues:
1) I believe the algorithm does not guarantee that the maximal subset with voting power < 1/3 is computed (as at some point those removed may be actually part of the maximal subset)
2) I ran a similar more complex but exact algorithm on the cosmos hub validators and since block 1 it can find a subset with total voting power = 1/3 - epsilon, with typical epsilon = 1. This means that the smallest change in voting power will cause (in theory) the "bad" voting power to be >= 1/3rd of nh.V. 
In hindsight this is expected with enough number of validators with wealth distribution for voting powers.
This brings us back to computing all subsets which is impractical (O(2**n)).
So in the end some form of periodic checkpoints + linear validation is more practical...unless I am completely wrong.

Reply:
Ah yes, I see, thanks.